Document rewriting of TCP based probes (see istio 33734) (#10119)
istio/istio.io: content/en/docs/ops/configuration/mesh/app-health-check/index.md

---

[Kubernetes liveness and readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/)
describes several ways to configure liveness and readiness probes:

1. [Command](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-command)
1. [HTTP request](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request)
1. [TCP probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-tcp-liveness-probe)

> The command approach works with Istio regardless of whether or not mutual TLS is enabled.

Reviewer: ```suggestion
The command approach works with no changes required, but HTTP requests and TCP probes require Istio to make changes to the pod configuration.
```

> The HTTP request approach, on the other hand, requires special Istio configuration when mutual TLS is enabled.
> This is because the health check requests to the `liveness-http` service are sent by Kubelet,
> which does not have an Istio issued certificate. Therefore when mutual TLS is enabled,
> the health check requests will fail.

Reviewer: ```suggestion
When mutual TLS is enabled, HTTP requests to the `liveness-http` service are sent by the Kubelet,
which does not have an Istio-issued certificate. Therefore, the requests will fail.
```

Author: Overall your suggestions look good to me. But this one is a bit broken.

"**When mutual TLS is enabled**, HTTP requests to the `liveness-http` service **are sent by the Kubelet**" This is wrong: the requests are always sent by the Kubelet, but it is only a problem when mTLS is enabled.

How about:
```suggestion
The health check requests to the `liveness-http` service are sent by Kubelet.
This becomes a problem when mutual TLS is enabled, because the Kubelet does not have an Istio issued certificate.
Therefore the health check requests will fail.
```

Reviewer: LGTM

> Istio solves this problem by rewriting the application `PodSpec` readiness/liveness probe,
> so that the probe request is sent to the [sidecar agent](/docs/reference/commands/pilot-agent/).
> The sidecar agent then redirects the request to the application, strips the response body, only returning the response code.
> 

Reviewer: ```suggestion
```

> TCP probe based checks always need special handling, because:
> 
> - Istio will redirect any incoming traffic into the sidecar any TCP port will appear open.
> - For TCP probes the Kubelet simply checks, if some process is listening on the specified port.
> 
> As a consequence the probes would always succeed as long as the sidecar is running.
> Istio will therefore also rewrite the application `PodSpec` readiness/liveness probes.
> The sidecar agent will then do the port check while avoiding the traffic redirection.
> 
> The rewriting of all problematic probes is enabled by default in all built-in Istio
> [configuration profiles](/docs/setup/additional-setup/config-profiles/) but can be disabled as described below.

Reviewer: ```suggestion
TCP probe checks need special handling, because Istio redirects all incoming traffic into the sidecar, and so all TCP ports appear open.  The Kubelet simply checks if some process is listening on the specified port, and so the proble will always succeed as long as the sidecar is running.

Istio solves both these problems by rewriting the application `PodSpec` readiness/liveness probe,
so that the probe request is sent to the [sidecar agent](/docs/reference/commands/pilot-agent/).
For HTTP requests, the sidecar agent redirects the request to the application and strips the response body, only returning the response code.  For TCP probes, the sidecar agent will then do the port check while avoiding the traffic redirection.

The rewriting of problematic probes is enabled by default in all built-in Istio
[configuration profiles](/docs/setup/additional-setup/config-profiles/) but can be disabled as described below.
```

The health check requests to the `liveness-http` service are sent by Kubelet.
This becomes a problem when mutual TLS is enabled, because the Kubelet does not have an Istio issued certificate.
Therefore the health check requests will fail.

> TCP probe checks need special handling, because Istio redirects all incoming traffic into the sidecar, and so all TCP ports appear open.  The Kubelet simply checks if some process is listening on the specified port, and so the probe will always succeed as long as the sidecar is running.

Reviewer: I got 99 probles but a typo ain't one

Reviewer: Need a blank line between 28 and 29

> ## Liveness and readiness probes using the HTTP request approach

Reviewer: HTTP request/TCP probe?

Author: We changed it to "Liveness and readiness probes using the HTTP or TCP approach" not to make the heading even longer.

Hope thats fine for you.

Reviewer: 👍

> As stated previously, Istio uses probe rewrite to implement HTTP probes by default. You can disable this
> feature either for specific pods, or globally.

Reviewer: ```suggestion
As stated previously, Istio uses probe rewrite to implement HTTP/TCP probes by default. You can disable this
feature either for specific pods, or globally.
```

I can't leave a suggestion on a line you didn't edit, but please change the heading on L85 to say "the HTTP request/TCP probe" too
